src/v/config/property_constraints.h

```diff
@@ -0,0 +1,268 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "config/convert.h"
+#include "config/from_string_view.h"
+#include "config/property.h"
+#include "json/_include_first.h"
+#include "json/stringbuffer.h"
+#include "json/writer.h"
+#include "utils/type_traits.h"
+
+#include <seastar/core/sstring.hh>
+#include <seastar/util/noncopyable_function.hh>
+
+#include <fmt/format.h>
+#include <yaml-cpp/node/node.h>
+
+#include <memory>
+#include <string>
+
+namespace cluster {
+struct topic_configuration;
+} // namespace cluster
+
+namespace config {
+
+struct constraint_validator_t {
+    virtual bool validate(const cluster::topic_configuration&) const = 0;
+    virtual void clamp(cluster::topic_configuration&) const = 0;
+    virtual void encode_yaml(YAML::Node&) const = 0;
+    virtual void rjson_serialize(json::Writer<json::StringBuffer>&) const = 0;
+    virtual void print(std::ostream&) const = 0;
+    virtual ~constraint_validator_t() = default;
+};
+
+using constraint_enabled_t = ss::bool_class<struct constraint_enabled_tag>;
+
+std::shared_ptr<constraint_validator_t> constraint_validator_map(
+  ss::sstring& name,
+  std::optional<ss::sstring> min_str_opt,
+  std::optional<ss::sstring> max_str_opt,
+  constraint_enabled_t enabled);
+
+template<typename T>
+void rjson_serialize_integral(
+  json::Writer<json::StringBuffer>& w, const std::optional<T>& val) {
+    if constexpr (std::is_same_v<T, std::chrono::milliseconds>) {
+        w.Uint64(val->count());
+    } else if constexpr (std::is_same_v<T, uint64_t>) {
+        w.Uint64(*val);
+    } else if constexpr (std::is_same_v<T, uint32_t>) {
+        w.Uint(*val);
+    } else if constexpr (
+      std::is_same_v<T, int32_t> || std::is_same_v<T, int16_t>) {
+        w.Int(*val);
+    } else {
+        static_assert(
+          utils::unsupported_type<T>::value,
+          "Type name is not a valid integral");
+    }
+}
+
+template<typename T>
+struct constraint_validator_range : public constraint_validator_t {
+    using extractor_t =
+      typename ss::noncopyable_function<T(const cluster::topic_configuration&)>;
+    using clamper_t = typename ss::noncopyable_function<void(
+      cluster::topic_configuration&, T&)>;
+
+    std::optional<T> min;
+    std::optional<T> max;
+    extractor_t value_extractor;
+    clamper_t clamper;
+
+    constraint_validator_range(
+      std::optional<T> min,
+      std::optional<T> max,
+      extractor_t value_extractor,
+      clamper_t clamper)
+      : min{std::move(min)}
+      , max{std::move(max)}
+      , value_extractor{std::move(value_extractor)}
+      , clamper{std::move(clamper)} {}
```

Comment:
issue: you need to validate min/max aren't conflicting (min <= max at all times)

Reply:
I marked to include this check in the path for PUT /cluster_config, check methods are [here](https://github.com/redpanda-data/redpanda/pull/14084/commits/df30c5825a50730fd259258837df9f1b1f37fe87) and use is [here](https://github.com/redpanda-data/redpanda/pull/14084/commits/f329d6daeaefa8810a2af3f0686231f13631f47a). I can move them into this patch instead.

Reply:
Those checks should be here for sure (in case some other mechanism is used at some point to create these).